### PR TITLE
API removing model attribute from criterion

### PR DESCRIPTION
**Pull request by mathurinm**

### CODE REVIEW

File: sparse_ho/criterion.py [hunk 1]
```diff
@@ -20,50 +20,40 @@ class HeldOutMSE():
...
         self.quantity_to_warm_start = None
         self.val_test = None
         self.rmse = None
 
-    def get_v(self, mask, dense):
-        return 2 * (self.X_val[:, mask].T @ (
-            self.X_val[:, mask] @ dense - self.y_val)) / self.X_val.shape[0]
-
-    def value(self, mask, dense):
+    def get_mse_val(self, X, y, mask, dense):
```
Comment (mathurinm) [inline sparse_ho/criterion.py:45]: @QB3  for me get_value was better, a consistent name across all `Criterion`

File: sparse_ho/criterion.py [hunk 2]
```diff
@@ -79,75 +69,76 @@ def compute_rmse(self, mask, dense, beta_star):
         else:
             self.rmse = None
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
```
Comment (mathurinm) [inline sparse_ho/criterion.py:72]: I don't get the distinction between this and get_mse_val. What does it do ?

File: sparse_ho/criterion.py [hunk 3]
```diff
@@ -79,75 +69,76 @@ def compute_rmse(self, mask, dense, beta_star):
...
 class HeldOutLogistic():
     """Logistic loss on held out data
     """
 
-    def __init__(self, X_val, y_val, model, X_test=None, y_test=None):
+    def __init__(self, idx_train, idx_val, X_test=None, y_test=None):
         """
         Parameters
         ----------
         X_val : {ndarray, sparse matrix} of (n_samples, n_features)
```
Comment (mathurinm) [inline sparse_ho/criterion.py:113]: remove

File: sparse_ho/criterion.py [hunk 4]
```diff
@@ -79,75 +69,76 @@ def compute_rmse(self, mask, dense, beta_star):
...
-        v /= self.X_val.shape[0]
-        return v
-
-    def value(self, mask, dense):
-        val = np.sum(
-            np.log(1 + np.exp(-self.y_val * (self.X_val[:, mask] @ dense))))
-        val /= self.X_val.shape[0]
-        return val
+    @staticmethod
+    def value_outer_crit(X, y, mask, dense):
```
Comment (mathurinm) [inline sparse_ho/criterion.py:134]: not sure what it does, why it handles X and y being None, why criterion.get_value is not enough

File: sparse_ho/criterion.py [hunk 5]
```diff
@@ -279,18 +270,19 @@ def get_val_grad(
...
-        mask, dense = self.model.get_primal(mask, dense)
+        mask, dense = model.get_primal(mask, dense)
         val = self.value(mask, dense)
         self.value_test(mask, dense)
         self.compute_rmse(mask, dense, beta_star)
 
         return val, grad
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, log_alpha, tol=1e-3):
```
Comment (mathurinm) [inline sparse_ho/criterion.py:280]: pass X and y here

File: sparse_ho/criterion.py [hunk 6]
```diff
@@ -355,31 +334,20 @@ def __init__(self, X, y, model, sigma, finite_difference_step=None,
...
-        val -= self.X_val.shape[0] * self.sigma ** 2
+        val = norm(y - X[:, mask] @ dense) ** 2
+        val -= X.shape[0] * self.sigma ** 2
         val += 2 * self.sigma ** 2 * dof
 
         return val
 
-    def value_test(self, mask, dense):
-        # self.val_test = None
+    def value_test(self, X, y, mask, dense):
```
Comment (mathurinm) [inline sparse_ho/criterion.py:348]: to simplify I would remove all X_test, y_test, value_test in this PR, so that we see more clearly which functions do what

File: sparse_ho/tests/test_elastic.py [hunk 7]
```diff
@@ -13,61 +13,57 @@
...
-    n_features=n_features, n_times=1, n_active=n_active, rho=rho,
-    SNR=SNR, seed=2)
-X_val = csc_matrix(X_val)
-alpha_max = (X_train.T @ y_train).max() / n_samples
+X_s = csc_matrix(X)
+
+idx_train = np.arange(0, 50)
+idx_val = np.arange(50, 100)
+
+alpha_max = (X[idx_train, :].T @ y[idx_train]).max() / n_samples
```
Comment (mathurinm) [inline sparse_ho/tests/test_elastic.py:31]: / len(idx_train)   and use norm(ord='inf') (an abs is lacking here)

File: sparse_ho/tests/test_grad_search.py [hunk 8]
```diff
@@ -21,26 +21,16 @@
...
-    n_features=n_features, n_times=1, n_active=n_active, rho=rho,
-    SNR=SNR, seed=1)
-X_test_s = csc_matrix(X_test)
-
-X_val, y_val, beta_star, noise, sigma = get_synt_data(
-    dictionary_type="Toeplitz", n_samples=n_samples,
-    n_features=n_features, n_times=1, n_active=n_active, rho=rho,
-    SNR=SNR, seed=2)
-X_test_s = csc_matrix(X_test)
+X_train_s = csc_matrix(X)
```
Comment (mathurinm) [inline sparse_ho/tests/test_grad_search.py:28]: should be X_s I think

File: sparse_ho/tests/test_grad_search.py [hunk 9]
```diff
@@ -21,26 +21,16 @@
...
 
+idx_train = np.arange(0, 50)
+idx_val = np.arange(50, 100)
 
-alpha_max = (X_train.T @ y_train).max() / n_samples
+alpha_max = np.max(np.abs(X[idx_train, :].T @ y[idx_train])) / n_samples
```
Comment (mathurinm) [inline sparse_ho/tests/test_grad_search.py:33]: / len(idx_train)

File: sparse_ho/tests/test_grid_search.py [hunk 10]
```diff
@@ -15,26 +15,16 @@
...
 
-X_val, y_val, beta_star, noise, sigma = get_synt_data(
-    dictionary_type="Toeplitz", n_samples=n_samples,
-    n_features=n_features, n_times=1, n_active=n_active, rho=rho,
-    SNR=SNR, seed=2)
-X_test_s = csc_matrix(X_test)
-
-
-alpha_max = (X_train.T @ y_train).max() / n_samples
+alpha_max = np.max(np.abs(X[idx_train, :].T @ y[idx_train])) / n_samples
```
Comment (mathurinm) [inline sparse_ho/tests/test_grid_search.py:27]: / len(idx_train)

File: sparse_ho/tests/test_logreg.py [hunk 11]
```diff
@@ -15,77 +15,66 @@
...
-X_val_s = csc_matrix(X_val)
+idx_train = np.arange(0, 50)
+idx_val = np.arange(50, 100)
 
-y_train[y_train == 0.0] = -1.0
-y_val[y_val == 0.0] = -1.0
-
-alpha_max = np.max(np.abs(X_train.T @ (- y_train)))
-alpha_max /= (2 * n_samples)
+alpha_max = np.max(np.abs(X[idx_train, :].T @ y[idx_train])) / (2 * n_samples)
```
Comment (mathurinm) [inline sparse_ho/tests/test_logreg.py:30]: len(idx_train)

File: sparse_ho/tests/test_logreg.py [hunk 12]
```diff
@@ -15,77 +15,66 @@
...
 log_alpha = np.log(alpha)
 tol = 1e-16
 
 models = [
-    SparseLogreg(
-        X_train, y_train, max_iter=10000, estimator=None),
-    SparseLogreg(
-        X_train_s, y_train, max_iter=10000, estimator=None)
+    SparseLogreg(max_iter=10000, estimator=None),
+    SparseLogreg(max_iter=10000, estimator=None)
```
Comment (mathurinm) [inline sparse_ho/tests/test_logreg.py:37]: not sure this will be usefull now

File: sparse_ho/tests/test_logreg.py [hunk 13]
```diff
@@ -15,77 +15,66 @@
...
 
 
 def get_v(mask, dense):
-    return 2 * (X_val[:, mask].T @ (
-        X_val[:, mask] @ dense - y_val)) / X_val.shape[0]
+    return 2 * (X[np.ix_(idx_val, mask)].T @ (
+        X[np.ix_(idx_val, mask)] @ dense - y[idx_val])) / len(idx_val)
 
 
 @pytest.mark.parametrize('model', models)
```
Comment (mathurinm) [inline sparse_ho/tests/test_logreg.py:55]: parametrize in terms of X and Xs maybe ?

File: examples/plot_held_out_lasso.py [hunk 14]
```diff
@@ -28,29 +28,26 @@
 from sparse_ho.grid_search import grid_search
 from sklearn.datasets import make_regression
 
-from sklearn.model_selection import train_test_split
+from libsvmdata.datasets import fetch_libsvm
```
Comment (mathurinm) [inline examples/plot_held_out_lasso.py:31]: import from libsvmdata

File: sparse_ho/tests/test_svm.py [hunk 15]
```diff
@@ -14,71 +14,66 @@
...
+C = 0.01
 log_C = np.log(C)
 tol = 1e-16
 
 models = [
-    SVM(
-        X_train, y_train, log_C, max_iter=10000, tol=tol),
-    SVM(
-        X_train_s, y_train, log_C, max_iter=10000, tol=tol)
+    SVM(log_C, max_iter=10000, tol=tol),
```
Comment (mathurinm) [inline sparse_ho/tests/test_svm.py:39]: @QB3 I think this is no longer needed and you need to parametrize your test with X, Xs instead

File: sparse_ho/criterion.py (hunk 1 above)
Comment (QB3) [reply sparse_ho/criterion.py:45]: I am going for ```value_outer_crit``` in all criterions

File: sparse_ho/criterion.py (hunk 2 above)
Comment (QB3) [reply sparse_ho/criterion.py:72]: The ```get_val```  function is for 0 order méthods, it compute the coefficients beta (without Jacobians) and the value of the criterion, whereas  ```get_mse_val``` compute the mse given the coefficients beta.

File: sparse_ho/criterion.py (hunk 4 above)
Comment (QB3) [reply sparse_ho/criterion.py:123]: Because we still monitor the value on the test using this, and X_test can be None, this will be removed using a callback function

File: examples/plot_sparse_log_reg.py [hunk 16]
```diff
@@ -73,14 +74,13 @@
 t0 = time.time()
 
 estimator = LogisticRegression(
-    penalty='l1', fit_intercept=False, solver='saga', max_iter=max_iter)
-model = SparseLogreg(X_train, y_train, max_iter=max_iter, estimator=estimator)
-criterion = HeldOutLogistic(X_val, y_val, model)
+    penalty='l1', solver='saga', fit_intercept=False, max_iter=max_iter)
```
Comment (mathurinm) [inline examples/plot_sparse_log_reg.py:77]: celer does not support this solver, I am surprised this work. easy fix: import from sklearn.linear_model

File: sparse_ho/criterion.py [hunk 17]
```diff
@@ -20,178 +20,157 @@ class HeldOutMSE():
...
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
-        else:
-            self.rmse = None
+    def value_outer_crit(self, X, y, mask, dense):
+        """Compute the MSE on the validation set.
+        """
+        if X is None or y is None:
+            return None
+        return norm(y - X[:, mask] @ dense) ** 2 / len(self.idx_val)
```
Comment (mathurinm) [inline sparse_ho/criterion.py:53]: len(y) would be safer

so t this function, you pass full X_test and y_test, not full X and y, right ?

File: sparse_ho/criterion.py (hunk 2 above)
Comment (mathurinm) [reply sparse_ho/criterion.py:55]: for me it's name shouldn't be get_val if it does something else. what do you think ?

Comment (mathurinm) [reply sparse_ho/criterion.py:55]: also, I got confused between MSE in prediction (the name of the class) and MSE in estimation. I would rename get_mse to get_mse_estim ?

Anyway, this is not something that the end used wants (he does not have beta_star). Couldn't it be handled with a callback ?

File: sparse_ho/criterion.py [hunk 18]
```diff
@@ -20,178 +20,157 @@ class HeldOutMSE():
...
         self.dense0 = dense
         self.quantity_to_warm_start = quantity_to_warm_start
-        mask, dense = self.model.get_primal(mask, dense)
-        val = self.value(mask, dense)
-        self.value_test(mask, dense)
-        self.compute_rmse(mask, dense, beta_star)
-        if self.convexify:
-            val += self.gamma_convex + np.sum(np.exp(log_alpha) ** 2)
-            grad += 2 * self.gamma_convex * np.exp(log_alpha)
+        # # g
```
Comment (mathurinm) [inline sparse_ho/criterion.py:79]: remove

File: sparse_ho/criterion.py [hunk 19]
```diff
@@ -355,76 +309,70 @@ def __init__(self, X, y, model, sigma, finite_difference_step=None,
...
-        dof = ((self.X_val[:, mask2] @ dense2 -
-                self.X_val[:, mask] @ dense) @ self.delta)
+    def value_outer_crit(self, X, y, mask, dense, mask2, dense2):
+        dof = ((X[:, mask2] @ dense2 -
+                X[:, mask] @ dense) @ self.delta)
         dof /= self.epsilon
-        # compute the value of the sure
-        val = norm(self.y_val - self.X_val[:, mask] @ dense) ** 2
-        val -= self.X_val.shape[0] * self.sigma ** 2
+        # compute the value_outer_crit of the sure
```
Comment (mathurinm) [inline sparse_ho/criterion.py:316]: ctrl f gone wrong

File: sparse_ho/criterion.py (hunk 2 above)
Comment (QB3) [reply sparse_ho/criterion.py:55]: > Anyway, this is not something that the end used wants (he does not have beta_star). Couldn't it be handled with a callback ?

ye it will handle in a callback

File: sparse_ho/criterion.py (hunk 17 above)
Comment (QB3) [reply sparse_ho/criterion.py:53]: > so t this function, you pass full X_test and y_test, not full X and y, right ?

yes

**Comment by QB3** [comment]
We removed the dependence in X and y when creating Models and Criterion, paving the way for a more efficient cross validation (that will be done in a another PR).
I think this is ready for review @agramfort

File: sparse_ho/criterion.py [hunk 20]
```diff
@@ -20,178 +20,156 @@ class HeldOutMSE():
...
+            X[self.idx_train, :], y[self.idx_train], mask, dense)
+        val = self.value_outer_crit(
+            X[self.idx_val, :], y[self.idx_val], mask, dense)
+        # TODO put the following in a callback function
+        self.mse_test = self.value_outer_crit(
+            self.X_test, self.y_test, mask, dense)
         return val, grad
 
+    def proj_param(self, model, X, y, log_alpha):
+        return model.proj_param(
```
Comment (agramfort) [inline sparse_ho/criterion.py:89]: what does proj_param do?

File: sparse_ho/criterion.py [hunk 21]
```diff
@@ -201,76 +179,58 @@ class HeldOutSmoothedHinge():
...
+        if X is None or y is None:
+            return None
 
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
+        if issparse(X):
+            Xbeta_y = (X[:, mask].T).multiply(y).T @ dense
```
Comment (agramfort) [inline sparse_ho/criterion.py:211]:
```suggestion
            Xbeta_y = (X[:, mask].T @ y).T @ dense
```

File: sparse_ho/criterion.py [hunk 22]
```diff
@@ -355,76 +308,69 @@ def __init__(self, X, y, model, sigma, finite_difference_step=None,
...
-            self, log_alpha, get_beta_jac_v,
+            self, model, X, y, log_alpha, get_beta_jac_v,
             mask0=None, dense0=None, beta_star=None,
             jac0=None, max_iter=1000, tol=1e-3, compute_jac=True):
+        if not self.init_delta_epsilon:
+            self._init_delta_epsilon(X)
+
+        def v(mask, dense):
+            return (2 * X[:, mask].T @ (
+                    X[:, mask] @ dense - y -
```
Comment (agramfort) [inline sparse_ho/criterion.py:353]: X[:, mask] makes copies of X. If you can avoid calling twice X[:, mask] it will save you time.

File: sparse_ho/criterion.py [hunk 23]
```diff
@@ -355,76 +308,69 @@ def __init__(self, X, y, model, sigma, finite_difference_step=None,
...
+            X, y + self.epsilon * self.delta,
+            log_alpha, model, v2, mask0=self.mask02,
             dense0=self.dense02,
             quantity_to_warm_start=self.quantity_to_warm_start2,
             max_iter=max_iter, tol=tol, compute_jac=compute_jac,
             full_jac_v=True)
-        val = self.value(mask, dense, mask2, dense2)
-        self.value_test(mask, dense)
-        self.compute_rmse(mask, dense, beta_star)
+        val = self.value_outer_crit(X, y, mask, dense, mask2, dense2)
```
Comment (agramfort) [inline sparse_ho/criterion.py:373]:
```suggestion
        val = self.get_val_outer(X, y, mask, dense, mask2, dense2)
```

so we have : get_val, get_val_grad for inner and get_val_outer ?

File: sparse_ho/implicit_forward.py [hunk 24]
```diff
@@ -56,13 +56,13 @@ def get_beta_jac_fast_iterdiff(
     v = None
     _, r = model._init_beta_r(X, y, mask, dense)
     jac = get_only_jac(
-        model.reduce_X(mask), model.reduce_y(mask), r, reduce_alpha, model.sign(dense, log_alpha), v,
+        model.reduce_X(X, mask), model.reduce_y(y, mask), r, n_samples, reduce_alpha, model.sign(dense, log_alpha), v,
```
Comment (agramfort) [inline sparse_ho/implicit_forward.py:59]: `model.reduce_X(X, mask)` is just `X[:, mask]` no?

if so can we simplify?

File: sparse_ho/implicit_forward.py [hunk 25]
```diff
@@ -56,13 +56,13 @@ def get_beta_jac_fast_iterdiff(
...
         dbeta=dbeta0_new, niter_jac=niter_jac, tol_jac=tol_jac, model=model, mask=mask, dense=dense, verbose=verbose)
     return mask, dense, jac
 
 
 def get_only_jac(
-        Xs, y, r, alpha, sign_beta, v, dbeta=None, niter_jac=100, tol_jac=1e-4, model="lasso", mask=None, dense=None, verbose=False):
+        Xs, y, r, n_samples, alpha, sign_beta, v, dbeta=None, niter_jac=100, tol_jac=1e-4, model="lasso", mask=None, dense=None, verbose=False):
```
Comment (agramfort) [inline sparse_ho/implicit_forward.py:65]: you pass n_samples and you immedially overwrite it in line 66

File: sparse_ho/models.py [hunk 26]
```diff
@@ -196,15 +195,14 @@ def _update_only_jac_sparse(
...
 
-    # @staticmethod
-    def _get_jac_t_v(self, jac, mask, dense, alphas, v):
-        n_samples = self.X.shape[0]
+    @staticmethod
+    def _get_jac_t_v(X, y, jac, mask, dense, alphas, v, n_samples):
         return n_samples * alphas[mask] * np.sign(dense) @ jac
 
-    def proj_param(self, log_alpha):
+    def proj_param(self, X, y, log_alpha):
```
Comment (agramfort) [inline sparse_ho/models.py:202]: I don't get what this does

File: sparse_ho/models.py [hunk 27]
```diff
@@ -230,26 +228,29 @@ def _use_estimator(self, X, y, alpha, tol, max_iter):
...
+    @staticmethod
+    def reduce_y(y, mask):
+        return y
 
     def sign(self, x, log_alpha):
         return np.sign(x)
 
-    def get_primal(self, mask, dense):
+    def get_primal(self, X, y, mask, dense):
         return mask, dense
```
Comment (agramfort) [inline sparse_ho/models.py:243]: this is weird

File: sparse_ho/models.py [hunk 28]
```diff
@@ -230,26 +228,29 @@ def _use_estimator(self, X, y, alpha, tol, max_iter):
...
 
-    def get_jac_v(self, mask, dense, jac, v):
+    def get_jac_v(self, X, y, mask, dense, jac, v):
         return jac.T @ v(mask, dense)
 
-    def get_hessian(self, mask, dense, log_alpha):
-        hessian = self.X[:, mask].T @ self.X[:, mask]
+    @staticmethod
+    def get_hessian(X_train, y_train, mask, dense, log_alpha):
+        hessian = X_train[:, mask].T @ X_train[:, mask]
```
Comment (agramfort) [inline sparse_ho/models.py:250]:
```suggestion
        X_m = X_train[:, mask]
        hessian = X_m.T @ X_m
```

File: sparse_ho/models.py [hunk 29]
```diff
@@ -484,8 +481,9 @@ def get_L(X, is_sparse=False):
         else:
             return norm(X, axis=0) ** 2 / (X.shape[0])
 
-    def get_hessian(self, mask, dense, log_alpha):
-        hessian = self.X[:, mask].T @ self.X[:, mask]
+    @staticmethod
+    def get_hessian(X, y, mask, dense, log_alpha):
+        hessian = X[:, mask].T @ X[:, mask]
```
Comment (agramfort) [inline sparse_ho/models.py:486]: idem

**Review by agramfort — COMMENTED** [review]
here is a first batch of comments @QB3 

writing docstrings should help you also clarify the API

if you can explain it in simple words it's a good sign !

File: sparse_ho/criterion.py [hunk 30]
```diff
@@ -20,178 +20,156 @@ class HeldOutMSE():
     # cv as [(train, test)] ie directly the indices of the train
     # and test splits.
 
-    def __init__(self, X_val, y_val, model, convexify=False,
-                 gamma_convex=1e-2, X_test=None, y_test=None):
+    def __init__(
+            self, idx_train, idx_val, X_test=None, y_test=None,):
```
[the remark this answers is not in the document]
Comment (mathurinm) [reply sparse_ho/criterion.py:24]: these are completely unseen data, to monitor some left out loss. They will be removed in a subsequent PR (the Monitor class should handle them, not criterion)

File: sparse_ho/criterion.py [hunk 31]
```diff
@@ -20,178 +20,156 @@ class HeldOutMSE():
...
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
         # TODO add warm start
         mask, dense, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y, log_alpha, self.model, tol=tol, compute_jac=False)
-        self.value_test(mask, dense)
-        return self.value(mask, dense)
+            X[self.idx_train], y[self.idx_train], log_alpha, model, tol=tol, compute_jac=False)
+        self.get_mse_test(mask, dense)
```
[the remark this answers is not in the document]
Comment (mathurinm) [reply sparse_ho/criterion.py:59]: I got confused too. This function returns MSE in estimation, norm(beta_k - beta_star) ** 2.    It should also be handled by Monitor, not criterion, and will be removed.

File: sparse_ho/criterion.py (hunk 21 above)
Comment (mathurinm) [reply sparse_ho/criterion.py:211]: Actually it's to perform row-wise multiplication while preserving sparsit of the output, it's not vector matrix multiplication

File: sparse_ho/criterion.py (hunk 23 above)
Comment (mathurinm) [reply sparse_ho/criterion.py:373]: :+1: and even get_val_inner, get_val_outer to clarify

File: sparse_ho/criterion.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (QB3) [reply sparse_ho/criterion.py:24]: val and test have the same meaning, ```X_test``` with be handled properly in another PR:
- either using a callback function
- I just thought we could also handle it using an ```idx_test``` parameter

File: sparse_ho/criterion.py (hunk 20 above)
Comment (QB3) [reply sparse_ho/criterion.py:89]: it ensure that the parameters ```log_alpha``` does not take value absurdly large, typically we bound it between ```log_alpha_max``` and ```log_alpha_max - 8```

File: sparse_ho/criterion.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (mathurinm) [reply sparse_ho/criterion.py:24]: IMO the callback should take as only argument beta, as in scipy API. You would do in your script:
```
X_test, y_ test = ....
def callback(beta):
    return norm(y_test - X_test@ beta), norm(beta - beta_star)

monitor = Monitor(callback=callback)
```

File: sparse_ho/criterion.py (hunk 20 above)
Comment (mathurinm) [reply sparse_ho/criterion.py:89]: Shouldn't it be an attribute of model ?

File: sparse_ho/models.py (hunk 26 above)
Comment (QB3) [reply sparse_ho/models.py:202]: As before, it ensure that the parameters ```log_alpha``` does not take value absurdly large, typically we bound it between ```log_alpha_max``` and ```log_alpha_max - 8```

File: sparse_ho/models.py (hunk 27 above)
Comment (QB3) [reply sparse_ho/models.py:243]: I know, this is because we also take into the svm, once this PR is merged I will open a new one to either improve or remove the svm

File: sparse_ho/criterion.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (josephsalmon) [reply sparse_ho/criterion.py:59]: I would recommend using norm(beta_k - beta_star) ** 2 /norm(beta_star) ** 2 to get something one can compare when varying n_samples or n_features. WDYT?

File: sparse_ho/criterion.py (hunk 20 above)
Comment (QB3) [reply sparse_ho/criterion.py:89]: > Shouldn't it be an attribute of model ?

For simple models, yes. For more complex criterions, like the cross validation, or the multiclass it is convenient to handle the projection in the criterion and not in the model.

File: sparse_ho/criterion.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (QB3) [reply sparse_ho/criterion.py:59]: > I got confused too. This function returns MSE in estimation, norm(beta_k - beta_star) ** 2. It should also be handled by Monitor, not criterion, and will be removed.
 
> I would recommend using norm(beta_k - beta_star) ** 2 /norm(beta_star) ** 2 to get something one can compare when varying n_samples or n_features. WDYT?

Yes and yes, this will be for the next PR! :)

[the remark this answers is not in the document]
Comment (QB3) [reply sparse_ho/criterion.py:59]: > as the class is called HeldOutMSE can you use a more generic name that get_mse_test? self.eval? self.compute?

The current name of this function is ```value_outer_crit```

File: sparse_ho/criterion.py (hunk 23 above)
Comment (QB3) [reply sparse_ho/criterion.py:376]: > get_val, get_val_grad for inner and get_val_outer ?

No, ```get_val_grad``` computes the value and the gradient of the outer criterion given a value of the regularization parameter, ```get_val``` compute the value  of the outer criterion given the value of the regularization parameter.

```get_val_outer``` computes the value of the outer criterion, given the coefficients.

If you think of clearer names I am in!

File: sparse_ho/models.py (hunk 27 above)
Comment (mathurinm) [reply sparse_ho/models.py:243]: the name could be more explicit, get_primal_vector ? get_coefs ?

Comment (agramfort) [reply sparse_ho/models.py:243]: coefs are called beta in the API so get_beta ?

File: sparse_ho/models.py (hunk 26 above)
Comment (mathurinm) [reply sparse_ho/models.py:202]: proj regularizer ? clip regularizer ? This is costly to recompute alpha_max every time, couldn't we pass max_val, min_val and use np.clip ?

File: sparse_ho/criterion.py (hunk 23 above)
Comment (mathurinm) [reply sparse_ho/criterion.py]: for starters get_val_and_grad for me, since I read get_val_grad as get_value_gradient, i.e. get gradient of value (and not, get value and its gradient)

File: sparse_ho/models.py (hunk 26 above)
Comment (QB3) [reply sparse_ho/models.py:202]: I went for ```proj_hyperparam```

File: sparse_ho/criterion.py [hunk 32]
```diff
@@ -20,178 +20,160 @@ class HeldOutMSE():
...
+            return None
         else:
-            self.rmse = None
+            val = np.sum(
+                np.log(1 + np.exp(-y * (X[:, mask] @ dense))))
+            val /= X.shape[0]
+            return val
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
```
Comment (mathurinm) [inline sparse_ho/criterion.py:132]: I would rename this also. It's confusing (with get_val_outer), and it does something more than getting value.

File: sparse_ho/implicit.py [hunk 33]
```diff
@@ -44,10 +44,10 @@ def get_beta_jac_t_v_implicit(
         X_train, y_train, log_alpha, mask0=mask0, dense0=dense0,
         tol=tol, max_iter=max_iter, compute_jac=False, model=model)
 
-    mat_to_inv = model.get_hessian(mask, dense, log_alpha)
+    mat_to_inv = model.get_hessian(X_train, y_train, mask, dense, log_alpha)
     size_mat = mat_to_inv.shape[0]
 
-    maskp, densep = model.get_primal(mask, dense)
+    maskp, densep = model. get_beta(X_train, y_train, mask, dense)
```
Comment (mathurinm) [inline sparse_ho/implicit.py:50]: extra space

**Comment by QB3** [comment]
I think we good to go, good for you @agramfort ?

File: examples/plot_grad_search_cv.py [hunk 34]
```diff
@@ -12,117 +12,117 @@
 #
 # License: BSD (3-clause)
```
Comment (agramfort) [inline examples/plot_grad_search_cv.py:14]: should you revert the changes here?

File: sparse_ho/implicit_forward.py [hunk 35]
```diff
@@ -72,12 +69,6 @@ def get_only_jac(
 
     if dbeta is None:
         model._init_dbeta(n_features)
-        # if model == "lasso":
-        #     dbeta = np.zeros(n_features)
-        # if model == "mcp":
-        #     dbeta = np.zeros((n_features, 2))
-        # elif model == "wlasso":
-        #     dbeta = np.zeros((n_features, n_features))
```
Comment (agramfort) [inline sparse_ho/implicit_forward.py:80]: cleanup?

File: sparse_ho/implicit_forward.py [hunk 36]
```diff
@@ -93,20 +84,9 @@ def get_only_jac(
...
-        # m2 = tol_jac * np.sqrt(n_features) * n_samples * alpha * norm(v)
-        # crit = m1 <= m2
-        # print("m1 %.2f", m1)
-        # print("m2 %.2f", m2)
-        # print("m1 = %f" % norm(v @ (dbeta - dbeta_old)))
-        # print("tol_crit %f" % tol_crit)
-        # if norm(v @ (dbeta - dbeta_old)) < tol_crit:
-        # if norm((dbeta - dbeta_old)) < tol_jac * norm(dbeta):
-        # crit =
-        # print('jac obj', objs[-1])
```
Comment (agramfort) [inline sparse_ho/implicit_forward.py:107]: cleanup?

File: sparse_ho/implicit_forward.py [hunk 37]
```diff
@@ -93,20 +84,9 @@ def get_only_jac(
...
+            model.get_jac_obj(Xs, y, n_samples, sign_beta, dbeta, r, dr, alpha))
+
         if i > 1 and np.abs(objs[-2] - objs[-1]) < np.abs(objs[-1]) * tol_jac:
             break
-        # dbeta_old = dbeta.copy()
-        # dr_old = dr.copy()
```
Comment (agramfort) [inline sparse_ho/implicit_forward.py:111]: cleanup?

File: sparse_ho/tests/test_elastic.py (hunk 7 above)
Comment (agramfort) [reply sparse_ho/tests/test_elastic.py:31]: np.abs missing

File: sparse_ho/tests/test_grad_search_CV.py [hunk 38]
```diff
@@ -1,92 +1,92 @@
...
-    assert np.allclose(objs_grid_sk, monitor_grid.objs)
-
-
-if __name__ == '__main__':
-    test_cross_val_criterion()
+# import numpy as np
+# from scipy.sparse import csc_matrix
+# import sklearn
+# from sklearn.model_selection import KFold
+# from sklearn.linear_model import LassoCV
```
Comment (agramfort) [inline sparse_ho/tests/test_grad_search_CV.py:5]: put back?

File: sparse_ho/tests/test_svm.py [hunk 39]
```diff
@@ -152,7 +135,6 @@ def test_grad_search(model):
     #     np.array(monitor1.objs_test), np.array(monitor3.objs_test))
     assert not np.allclose(
         np.array(monitor1.times), np.array(monitor3.times))
-    # import ipdb; ipdb.set_trace()
```
Comment (agramfort) [inline sparse_ho/tests/test_svm.py:155]: cleanup?

File: examples/plot_grad_search_cv.py (hunk 34 above)
Comment (QB3) [reply examples/plot_grad_search_cv.py:14]: I planned to do the CV in a separated PR because this one is becoming too big

File: sparse_ho/implicit_forward.py (hunk 36 above)
Comment (QB3) [reply sparse_ho/implicit_forward.py:107]: This is very werid, my ```implicit_forward.py``` does not have more than 100lines:
https://github.com/QB3/sparse-ho/blob/1313e032f28b7e0f85649367b51dbbe829249813/sparse_ho/implicit_forward.py

**Comment by agramfort** [comment]
@QB3 merge when CIs are green
